Source/Core/GoogleEarthEnterpriseTerrainProvider.js

```diff
@@ -0,0 +1,587 @@
+            return new HeightmapTerrainData({
+                buffer : new Uint8Array(16 * 16),
+                width : 16,
+                height : 16
+            });
+        } else if (terrainState === TerrainState.NONE) {
+            // Already have info and there isn't any terrain here
+            return when.reject(new RuntimeError('Terrain tile doesn\'t exist'));
+        }
+
+        // Figure out where we are getting the terrain and what version
+        var parentInfo;
+        var q = quadKey;
+        var terrainVersion = -1;
+        if (terrainState !== TerrainState.NONE) {
+            switch (terrainState) {
+                case TerrainState.SELF: // We have terrain and have retrieved it before
+                    terrainVersion = info.terrainVersion;
+                    break;
+                case TerrainState.PARENT: // We have terrain in our parent
+                    q = q.substring(0, q.length - 1);
+                    parentInfo = metadata.getTileInformationFromQuadKey(q);
+                    terrainVersion = parentInfo.terrainVersion;
+                    break;
+                case TerrainState.UNKNOWN: // We haven't tried to retrieve terrain yet
+                    if (info.hasTerrain()) {
+                        terrainVersion = info.terrainVersion; // We should have terrain
+                    } else {
+                        q = q.substring(0, q.length - 1);
+                        parentInfo = metadata.getTileInformationFromQuadKey(q);
+                        if (defined(parentInfo) && parentInfo.hasTerrain()) {
+                            terrainVersion = parentInfo.terrainVersion; // Try checking in the parent
+                        }
+                    }
+                    break;
+            }
+        }
+
+        // We can't figure out where to get the terrain
+        if (terrainVersion < 0) {
+            return when.reject(new RuntimeError('Terrain tile doesn\'t exist'));
+        }
+
+        // Load that terrain
+        var terrainPromises = this._terrainPromises;
+        var url = buildTerrainUrl(this, q, terrainVersion);
+        var promise;
+        if (defined(terrainPromises[q])) { // Already being loaded possibly from another child, so return existing promise
+            promise = terrainPromises[q];
+        } else { // Create new request for terrain
+            var requestPromise;
+            throttleRequests = defaultValue(throttleRequests, true);
+            if (throttleRequests) {
+                requestPromise = throttleRequestByServer(url, loadArrayBuffer);
+                if (!defined(requestPromise)) {
+                    return undefined; // Throttled
+                }
+            } else {
+                requestPromise = loadArrayBuffer(url);
+            }
+
+            promise = requestPromise
+                .then(function(terrain) {
+                    if (defined(terrain)) {
+                        return taskProcessor.scheduleTask({
```

Comment:
This is not ideal because it subverts the terrain system's ability to prioritize the work it does.  The way this is written, when a tile request comes back, it will always do work in the web worker, even if the user has since moved the camera and the tile is no longer needed.

A better approach would be to return a `TerrainData` representing the raw downloaded data, and only decode it on demand in `createMesh` and similar methods.

The rule of thumb for good terrain loading performance is: do as little as possible in promise resolutions, because they always happen, even if they're not needed.

Up to you if you think this is worth the trouble, of course.